whois-commons/src/main/java/net/ripe/db/whois/common/rpki/RpkiService.java

```diff
@@ -20,24 +23,35 @@ public class RpkiService {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(RpkiService.class);
 
-    private final NestedIntervalMap<Ipv4Resource, Set<Roa>> ipv4Tree = new NestedIntervalMap<>();
-    private final NestedIntervalMap<Ipv6Resource, Set<Roa>> ipv6Tree = new NestedIntervalMap<>();
+    final IntervalMap<Ipv4Resource, Set<Roa>> ipv4Tree = SynchronizedIntervalMap.synchronizedMap(new NestedIntervalMap<Ipv4Resource, Set<Roa>>());
+    final IntervalMap<Ipv6Resource, Set<Roa>> ipv6Tree = SynchronizedIntervalMap.synchronizedMap(new NestedIntervalMap<Ipv6Resource, Set<Roa>>());
+
+    private final RpkiDataProvider rpkiDataProvider;
 
     public RpkiService(final RpkiDataProvider rpkiDataProvider) {
+        this.rpkiDataProvider = rpkiDataProvider;
+        loadRoas();
+    }
+
+    @Scheduled(cron = "* /15 * * * *")
+    private void loadRoas() {
         final List<Roa> loadedRoas = rpkiDataProvider.loadRoas();
         if (loadedRoas != null && !loadedRoas.isEmpty()){
             final List<Roa> roas = loadedRoas.stream()
                     .filter(roa -> roa.getTrustAnchor() != TrustAnchor.UNSUPPORTED)
                     .collect(Collectors.toList());
 
+            ipv4Tree.clear();
```

Comment:
Asynchronous queries to the tree will find nothing after `clear()`. Maybe generate *new* ipv4Tree and ipv6Tree instances from scratch, and then assign them to the fields so there is no gap. Similar to `IpTreeCacheManager`

Reply:
Done